### DISCUSSION TITLE
Code generation for a Simulink model, that calls ipopt

### DISCUSSION THREAD

**Discussion by maskros-m** [casadi/casadi#3894] [Q&A]
Dear developers and users,

I have an optimization problem, written with Casadi and generated C code for the IPOPT solver. The code is integrated to a wrapper in C, sucessfully compiled to .mex file and called in a Simulink model via an S-function. So the simulation is working. Now I'm looking at generating C code for this Simulink model via Embedded Coder. While generating the code for this simulink model, I encountered the issue, that some functions in IPOPT are not defined as seen in the following picture.
![nocall_ipopt](https://github.com/user-attachments/assets/b4cdb7fd-4773-4290-bcef-2c897dfccbfd)

So, I have been thinking that the IPOPT library is missing. I have tried to include file ipopt.lib (copied from Casadi repo) in the Simulink model and encountered another issue, it appears reading 'ipopt.lib' causes issue

> ### Creating standalone executable "..\EOPT_C_ipopt_test.exe" ... 
> 	link /RELEASE  /INCREMENTAL:NO /NOLOGO kernel32.lib ws2_32.lib mswsock.lib advapi32.lib -out:..\EOPT_C_ipopt_test.exe @EOPT_C_ipopt_test.rsp C:\Users\ngocngu4\Documents\GitHub\mi_ptceu_pcpc\.venv\Lib\site-packages\casadi\ipopt.lib   
>    Bibliothek "..\EOPT_C_ipopt_test.lib" und Objekt "..\EOPT_C_ipopt_test.exp" werden erstellt. 
> EOPT_C_ipopt_test.obj : error LNK2019: Verweis auf nicht aufgel├Âstes externes Symbol "EOPT_ipopt_call" in Funktion "EOPT_C_ipopt_test_step". 
> ..\EOPT_C_ipopt_test.exe : fatal error LNK1120: 1 nicht aufgel├Âste Externe 
> NMAKE : fatal error U1077: "link /RELEASE  /INCREMENTAL:NO /NOLOGO kernel32.lib ws2_32.lib mswsock.lib advapi32.lib -out:..\EOPT_C_ipopt_test.exe @EOPT_C_ipopt_test.rsp C:\Users\ngocngu4\Documents\GitHub\mi_ptceu_pcpc\.venv\Lib\site-packages\casadi\ipopt.lib  ": R³ckgabe-Code "0x460" 
> Stop. 
>  
> C:\Users\ngocngu4\Documents\GitHub\mi_ptceu_pcpc\matlab\mdl\Cmpnts\MesoPlanner\eopt_standalone\EOPT_C_ipopt_test_grt_rtw>echo The make command returned an error of 2  
> The make command returned an error of 2 
>  
> C:\Users\ngocngu4\Documents\GitHub\mi_ptceu_pcpc\matlab\mdl\Cmpnts\MesoPlanner\eopt_standalone\EOPT_C_ipopt_test_grt_rtw>exit /B 1  
> ### Build procedure for EOPT_C_ipopt_test aborted due to an error.
> 

I would like to ask whether anyone, who has experience dealing IPOPT library with Embedded Coder or had similar problem, could share with me opinions/show me some hints to resolve it. Thank you in advance!

Best regards,
Anh

**Comment by phmandl** [comment]
have you included the libraries and header files in the settings of the simulink model under code generation -> custom code? Similarly, how you tell the mex command with -lipopt, you have to tell the simulink coder where to find the libraries.

**Reply by maskros-m** [reply]
Hi Philipp,

Indeed **ipopt.lib** is included in the field **Code Generation --> Custom Code --> Libraries** , as the error also shows the complete path to library file. But it seems reading anything in this field causes error. I have tried also to put it in the same folder with the Simulink Model, but the same error is persistent. 

Under **Code Generation --> Custom Code -->Include directories**, I also include the path to casadi and casadi\include. But not sure whether it corresponds to setting **['-L',lib_folder]** in mex command

Setting -lipopt in mex command amounts to calling ipopt.lib file in Windows?

**Reply by phmandl** [reply]
Could you make a simple example project of what you are trying to do and share it so i can test it on my hardware?

**Reply by maskros-m** [reply]
I left the necessary files in this git repo. The file ipopt.lib hasnt been included yet under Custom Code, but code generation doesn't work either in that case. Maybe you could find something. Thanks in advance!
https://github.com/maskros-m/Testfolder
